### ISSUE TITLE
Inconsistent `@MainActor` semantics on top level

### ISSUE THREAD

**Issue by stevapple** [apple/swift#59272]
**Describe the bug**

Top level codes are implicitly `@MainActor`, and the compiler will emit a warning if you use `await` on a `MainActor`-isolated function/variable. However, if you drop `await`, it will report an error saying `main actor-isolated … can not be referenced from a non-isolated context`.

**To Reproduce**

```swift
// main.swift
enum A {
  @MainActor
  static var prop: Int { 1 }
}
_ = A.prop
```

```swift
// main.swift
enum A {
  @MainActor
  static var prop: Int { 1 }
}
_ = await A.prop
```

**Expected behavior**
Either of these should compile without warning.

**Screenshots**

**Environment (please complete the following information):**
 - OS: macOS 12.4
 - Tag: swift-DEVELOPMENT-SNAPSHOT-2022-06-02-a

**Additional context**

SE-0343 requires implicit `await` to change top-level into asynchronous context, so it doesn’t compile without `await`. But the top-level context is main actor-isolated, so `await` shouldn’t be necessary. If this violates SE-0343 laws, we probably have to amend it.

**Comment by stevapple** [comment]
cc @etcwilde

**Comment by etcwilde** [comment]
This is correct behavior as per SE-0343. The changes in proposal only apply when the top-level is either an asynchronous context through the use of an explicit `await`/`async let` in the top-level, or you have passed `-warn-concurrency` to the compiler invocation.

This decision was made to avoid source-breaks of existing code when variables declared in top-level code are used as global variables from other files in the module. In this case, you would see an error like `var 'foo' isolated to global actor 'MainActor' can not be referenced from this synchronous context`. We accepted the source break when the top-level is transitioned to an asynchronous context because it requires an explicit action from the programmer, while untouched code will still compile and have the same behavior as before.

I am planning on adjusting the model everywhere in the fix for https://github.com/apple/swift/issues/57991, but that will be a separate proposal and something for Swift 6.

**Comment by stevapple** [comment]
I noticed that this is the correct behavior defined in SE-0343, but this issue raises another case where we may need explicit `async` entry point in addition to `async let` and `await`.

**Comment by etcwilde** [comment]
For now, we have the asynchronous main function
```swift
@main struct MainType {
  static func main() async {
    ...
   }
}
```
(you'll need to pass `-parse-as-library` to the compiler invocation if you only have a main.swift.)

That function is implicitly `MainActor` protected, so it will work for your needs.
Alternatively, if your code is concurrency safe, you can pass `-warn-concurrency` to the compiler invocation and that should also transition the top-level context to a `MainActor` protected context as well.
